libsbom/spdx.py

```diff
@@ -0,0 +1,220 @@
+import datetime
+import uuid
+
+from spdx_tools.spdx.writer.json     import json_writer
+from spdx_tools.spdx.writer.tagvalue import tagvalue_writer
+from spdx_tools.spdx.writer.xml      import xml_writer
+from spdx_tools.spdx.writer.yaml     import yaml_writer
+from spdx_tools.spdx.model import (
+    Actor,
+    ActorType,
+    Annotation,
+    AnnotationType,
+    Checksum,
+    ChecksumAlgorithm,
+    CreationInfo,
+    Document,
+    ExternalPackageRef,
+    ExternalPackageRefCategory,
+    Package,
+    Relationship,
+    RelationshipType
+)
+from spdx_tools.spdx.model.spdx_none import SpdxNone
+
+from . import constants
+from version import __version__
+
+writers = {
+    'json':     json_writer,
+    'tagvalue': tagvalue_writer,
+    'xml':      xml_writer,
+    'yaml':     yaml_writer,
+}
+
+def component_get_property(component: dict, property_name: str) -> str:
+    for prop in component["properties"]:
+        if prop["name"] == property_name:
+            return prop["value"]
+
+    return None
+
+
+def make_cpe_ref(cpe: str) -> ExternalPackageRef:
+    return ExternalPackageRef(ExternalPackageRefCategory.SECURITY,
+                              "cpe23Type", cpe)
+
+
+def make_purl_ref(purl: str) -> ExternalPackageRef:
+    return ExternalPackageRef(ExternalPackageRefCategory.PACKAGE_MANAGER,
+                              "purl", purl)
+
+
+def make_annotation(spdxid: str, content: str) -> Annotation:
+    this_tool = Actor(actor_type=ActorType.TOOL,
+                      name=f"alma-sbom {__version__}")
+
+    return Annotation(spdx_id=spdxid,
+                      annotation_type=AnnotationType.OTHER,
+                      annotator=this_tool,
+                      annotation_date=datetime.datetime.now(),
+                      annotation_comment=content)
+
+
+def make_checksum(algo: str, value: str) -> Checksum:
+    algo_map = {
+        "SHA-256": ChecksumAlgorithm.SHA256
+    }
+
+    if not algo in algo_map:
+        raise ValueError(f"Algorithm {algo} is not supported")
+
+    return Checksum(algo_map[algo], value)
+
+
+def buildtime_to_datetime(buildtime: str) -> datetime.datetime:
+    """Parse buildtime without nanoseconds into a datetime object.
+
+    The build timestamp is stored in nanosecond-precision in CAS, but
+    datetime.fromisoformat() on Python 3.10 or older is unable to parse
+    timestamps with sub-microsecond precision.
+    """
+    return datetime.datetime.fromisoformat(buildtime[:-4])
+
+
+def component_get_buildtime(component: dict) -> datetime.datetime:
+    buildtime = component_get_property(component, "almalinux:package:timestamp")
+
+    # Components in build SBOMs do not have timestamps
+    try:
+        return buildtime_to_datetime(buildtime)
+    except TypeError:
+        return None
+
+
+def build_get_timestamp(build: dict) -> datetime.datetime:
+    buildtime = component_get_property(build, "almalinux:albs:build:timestamp")
+
+    try:
+        return buildtime_to_datetime(buildtime)
```

Comment:
I'd say we do not need this here, as we get:
```
{
    "name": "almalinux:albs:build:timestamp",
    "value": "2023-07-25T10:13:50.445229"
}
```
and `datetime.datetime.fromisoformat` will be happy about it

Reply:
Oops, I didn't notice the two timestamps have different formats. I just fixed the commit.